core/object/class_db.cpp

```diff
@@ -491,7 +491,7 @@ StringName ClassDB::get_compatibility_class(const StringName &p_class) {
 	return StringName();
 }
 
-Object *ClassDB::_instantiate_internal(const StringName &p_class, bool p_require_real_class) {
+Object *ClassDB::_instantiate_internal(const StringName &p_class, bool p_require_real_class, bool p_skip_post_initialize) {
```

Comment:
Should we print a warning in this function if someone tries to create an extension class while `p_skip_post_initialize` is true? Because with extension classes, it won't actually skip post initialize, only native classes.

This may actually be a weakness of this approach versus PR https://github.com/godotengine/godot/pull/91018 - that one could allow also creating extension classes that aren't postinitialized, whereas this one has the inconsistency that it can't do that.

Although, I'm not sure what the use case would be for creating extension classes without calling postinitialize? Maybe when we eventually support extension classes extending classes from other extensions we may need that? In that case, the class furthest down in the heirarchy may need to create an instance of its parent class, but want to defer postinitialize until after it's finished initializing. But I'm not 100% sure we'll need that.

Reply:
> Should we print a warning in this function if someone tries to create an extension class while `p_skip_post_initialize` is true? Because with extension classes, it won't actually skip post initialize, only native classes.

I think this is redundant, `ClassDB::_instantiate_internal` is private function. it would not be used directly.



> This may actually be a weakness of this approach versus PR #91018 - that one could allow also creating extension classes that aren't postinitialized, whereas this one has the inconsistency that it can't do that.

Both this and [#91018](https://github.com/godotengine/godot/pull/91018) are can't create an extension class without postinitialization. 



> Although, I'm not sure what the use case would be for creating extension classes without calling postinitialize? Maybe when we eventually support extension classes extending classes from other extensions we may need that? In that case, the class furthest down in the heirarchy may need to create an instance of its parent class, but want to defer postinitialize until after it's finished initializing. But I'm not 100% sure we'll need that.

How to create an extension class is depend on `extension->create_instance`. If we need to create extension classes without postinitialization, we need a new `GDExtensionClassCreationInfo4` to support this.

If it support extend/inherit an extension class from a GDExtension library to another library through ClassDB (expecially cross languages binding), I think create extension classes without postinitialization is required.

But for now, this goal seems too far away and the demand is not clear.
I think we should not handle it in this pr, in other words, we can add this feature in another pr if we ensure that we need it.

Reply:
> I think this is redundant, ClassDB::_instantiate_internal is private function. it would not be used directly.

Sure, but `ClassDB::instantiate_without_postinitialization()` provides a way to call it, and it's inconsistent that it won't actually skip `NOTIFICATION_POSTINITIALIZE` if the class is an extension class. Inside of `ClassDB::_instantiate_internal()` it's fairly easy to check if it's an extension class, so that'd be a good place to put a check and a warning message.

> How to create an extension class is depend on extension->create_instance. If we need to create extension classes without postinitialization, we need a new GDExtensionClassCreationInfo4 to support this.

Yes. I'm suggesting that if we go with PR https://github.com/godotengine/godot/pull/91018, we should make a `GDExtensionClassCreationInfo4` that changes `create_instance_func` to a new signature which accepts a bool, for example:

```
typedef GDExtensionObjectPtr (*GDExtensionClassCreateInstance2)(void *p_class_userdata, bool p_do_postinitialize);
```

Then we could make `ClassDB::instanstiate_without_postinitialization()` consistent for both native and extension classes.

Reply:
> But for now, this goal seems too far away and the demand is not clear.

Also, I think this goal isn't all that far away. With the current effort to port C# support to GDExtension, it seems likely that folks will want to have C# classes descend from classes provided by other GDExtensions. I suspect we'll be digging into this for Godot 4.4.